### PR TITLE
better heuristic distinction of .d files

### PR DESCRIPTION
**Pull request by MartinNowak**
- properly recongnize dtrace probes
- recongnize \ in Makefile paths
- recongnize single line `file.ext : dep.ext` make targets
- recognize D module, import, function, and unittest declarations
- add more representative D samples

`rake benchmark:compare`
D changed from 31.2% to 28.1%
DTrace changed from 33.5% to 32.5%
Makefile changed from 35.3% to 39.4%

diff of `rake benchmark:generate`
https://gist.github.com/MartinNowak/6fba5aed9401dba4d4bfd7969a168c61 (URLs relative to github.com/)
- no misclassifications except for 2 /etc/init.d shell scripts

see https://gist.github.com/MartinNowak/fda24fdef64f2dbb05c5a5ceabf22bd3 for the scraper used to get test corpora for D, DTrace, and Makefile with `extension:d`.
- corpus had ~276K lines (~22MiB) from 3K files
- had to filter out lots of .d files from the binutils testsuite (apparently references file diff tests)

### CODE REVIEW

File: samples/D/arrayops.d [hunk 1]
```diff
@@ -0,0 +1,187 @@
+/**
+ * Benchmark for array ops.
+ *
+ * Copyright: Copyright Martin Nowak 2016 -.
+ * License:   $(LINK2 http://www.boost.org/LICENSE_1_0.txt, Boost License 1.0)
+ * Authors:   Martin Nowak
+ */
```
Comment (MartinNowak) [inline samples/D/arrayops.d]: This file being detected as Makefile brought me here.

**Comment by wilzbach** [comment]
Pinging this as I think the false-positive rate is a lot higher!
@github/linguist

https://github.com/libmir/mir/pull/281/files (most example scripts in `examples/flex_plot`)
https://gist.github.com/wilzbach/ed5033e5be08b09dc181972a4f56fb7c
https://gist.github.com/wilzbach/f2e5fc48e76c92fef7eb4f8b55b80e4b
https://gist.github.com/wilzbach/afbc2fffd770bfa98c6d353904df687a
... 

and a lot more

**Comment by larsbrinkhoff** [comment]
Thank you, @MartinNowak, this looks good!  Could you please help by explaining the regexp used to match function definitions?

**Comment by wilzbach** [comment]
> Thank you, @MartinNowak, this looks good! Could you please help by explaining the reasoning behind the regexp used to match function definitions?

From above:

> additionally match import statements, function definitions, and unittests all of those should not match DTrace or Makefiles

**Comment by larsbrinkhoff** [comment]
@wilzbach I'm asking for an explanation of how `\w+\s+\w+\s*\(.*\)(?:\(.*\))?\s*{[^}]*}` matches D function definitions, and why it doesn't match DTrace files.

**Comment by wilzbach** [comment]
On August 23, 2016 8:38:48 AM GMT+02:00, Lars Brinkhoff <notifications@github.com> wrote:
>@wilzbach I'm asking for an explanation of how
>`\w+\s+\w+\s*\(.*\)(?:\(.*\))?\s*{[^}]*}` matches D function
>definitions, and why it doesn't match DTrace files.

Because that's the pattern for a function in D and I highly doubt DTrace has this? 
The regex above matches this pattern:

```
<ReturnType> <FunctionName> ( <FunctionParameters>) {...} 
```

D allows templates, hence the optional look ahead:

```
<ReturnType> <FunctionName> ( <TemplateParameters>)(<FunctionParameters>){ ... } 
```

Or with a real function:

```d
auto myFun(T)(T paramA)  {
    return paramA + 1;
} 
```

For more information, you may also have a look at the spec: http://dlang.org/spec/function.html

**Comment by larsbrinkhoff** [comment]
Thank you @wilzbach!

@arfon This looks fine to me.  Possibly verify with a `.d` corpus?

**Comment by wilzbach** [comment]
> @arfon This looks fine to me. Possibly verify with a .d corpus?

Didn't @MartinNowak already add multiple files?

In any case a good repository for idiomatic D is of course it's standard library [Phobos] (https://github.com/dlang/phobos)  (about 200 kLoC iirc) and there are many more here [on  GitHub](https://github.com/search?utf8=✓&q=language%3AD&type=Repositories&ref=advsearch&l=D&l=) (the popularity and presence of D is rising slowly, but steadily)

**Comment by larsbrinkhoff** [comment]
@wilzbach I'm alluding to this phrase from CONTRIBUTING.md:

> Test the performance of the Bayesian classifier with a relatively large number (1000s) of sample `.yourextension` files. (ping @arfon or @bkeepers to help with this) to ensure we're not misclassifying files.

**Comment by wilzbach** [comment]
> @wilzbach I'm alluding to this phrase from CONTRIBUTING.md:

Oh I didn't read the guide as I am just following this PR. I could also submit more files as a follow-up to this PR?

**Comment by arfon** [comment]
> @arfon This looks fine to me. Possibly verify with a .d corpus?

Sure thing @larsbrinkhoff. Do we have a search query I can use to grab these files? I don't see one in this thread.

**Comment by wilzbach** [comment]
> Sure thing @larsbrinkhoff. Do we have a search query I can use to grab these files? I don't see one in this thread.

I am pretty sure you are more familiar with your search query, but I would start by fetching some of the most starred repos with language D and filter for all files with the suffix ".d"

https://github.com/search?utf8=%E2%9C%93&q=language%3Ad&type=Repositories&ref=searchresults

**Comment by arfon** [comment]
Thanks @wilzbach. I ended up going with [this search string](https://github.com/search?utf8=%E2%9C%93&q=NOT+asdkjasdkjahsbk+language%3Ad&type=Code&ref=searchresults) to return file results (we need to use the code search rather than the repository search here). @larsbrinkhoff - I just sent you a corpus - would you be able to take a look?

**Comment by wilzbach** [comment]
> Thanks @wilzbach. I ended up going with this search string to return file results (we need to use the code search rather than the repository search here). @larsbrinkhoff - I just sent you a corpus - would you be able to take a look?

Oh, one thing that you might be careful with is that even though D1 has been officially deprecated in 2010, there's still some code using the old language grammar and D2 constantly keeps improving, so I don't know how sensitive your classifier is, but you might want to use a [minimum date](https://github.com/search?utf8=%E2%9C%93&q=NOT+asdkjasdkjahsbk+language%3Ad+pushed%3A%3E2016-01-01&type=Code&ref=searchresults).

Secondly I noticed that your recognition of `language:d` isn't perfect, for example the first four files are not D source files:
https://github.com/sjnewbury/multilib-overlay/blob/80c9fd20cfd05481ac19edcadd56ad5e578a8930/net-wireless/bluez/files/4.18/bluetooth-init.d, but I don't how to fix this :/

**Comment by larsbrinkhoff** [comment]
> @larsbrinkhoff - I just sent you a corpus - would you be able to take a look?

Thanks.  It may take a while before I find a time slot for this.

If @MartinNowak or @wilzbach (or anyone else) feel this is urgent, I can provide instructions for how to evalutate the corpous.

**Comment by wilzbach** [comment]
> If @MartinNowak or @wilzbach (or anyone else) feel this is urgent, I can provide instructions for how to evalutate the corpous

Yes I consider this as urgent - how can I help? :)

**Comment by larsbrinkhoff** [comment]
@wilzbach You'll have to install Linguist locally as per the instructions [here](https://github.com/github/linguist/blob/master/CONTRIBUTING.md#testing).  Then I can send you a zip file with lots of `.d` files, and you can run linguist on them to ensure that this PR does the right thing.

**Comment by wilzbach** [comment]
> @wilzbach You'll have to install Linguist locally as per the instructions here. Then I can send you a zip file with lots of .d files, and you can run linguist on them to ensure that this PR does the right thing

Sure, that's easy (currently on my commute & can do this tonight) - did you see my comment about the search query finding also non D language files?

**Comment by larsbrinkhoff** [comment]
@wilzbach Great, can I send the file to you gmail address?

Right, `language:d` isn't perfect.  The improved heuristic should make it better.

**Comment by wilzbach** [comment]
> @wilzbach Great, can I send the file to you gmail address?

Yes, please do :)

**Comment by larsbrinkhoff** [comment]
You may get some files that are not D, but that's not a bad thing.

**Comment by arfon** [comment]
Sure thing @MartinNowak:

[`D`](https://github.com/github/linguist/files/490825/samples-59.zip)
[`DTrace`](https://github.com/github/linguist/files/490826/samples-60.zip)
[`Makefile`](https://github.com/github/linguist/files/490827/samples-61.zip)

**Comment by MartinNowak** [comment]
Thanks @arfon, unfortunately the zip files are empty.

**Comment by arfon** [comment]
> Thanks @arfon, unfortunately the zip files are empty.

Oh wow. Sorry about that @MartinNowak. Totally should have checked...

**Comment by MartinNowak** [comment]
Did you have time to refetch them @arfon. It seems the rules in this PR could use a bit more tweaking.

**Comment by arfon** [comment]
> Did you have time to refetch them @arfon. It seems the rules in this PR could use a bit more tweaking.

I did not I'm afraid. It's slightly more complex than I first though to fix this. We've been re-working how we run all of our asynchronous jobs in our platform and currently the admin tool I have to run these jobs is broken 😞. I'll report back here once this is fixed and we can pull the sample files again.

**Comment by MartinNowak** [comment]
Any other idea? I can try to make my script work a bit better, but using whole repos suffers badly from misclassification.

**Comment by MartinNowak** [comment]
Any workarounds @arfon?

**Comment by arfon** [comment]
> Any workarounds @arfon?

Not really sorry and I'm afraid since this Pull Request was opened I've left GitHub so we'll need to get another Linguist maintainer who's on GitHub staff to assist here.

@MikeMcQuaid @lildude @brandonblack - there's an area of staff tools `devtools/samples` I think that allows you to run a search query and it batches up the first ~1000 files. These are useful for testing new heuristics. Unfortunately the jobs don't seem to be running any longer and I didn't manage to figure out what was going on before I left. Any chance one of you could take a look?

**Comment by wilzbach** [comment]
> Not really sorry and I'm afraid since this Pull Request was opened I've left GitHub so we'll need to get another Linguist maintainer who's on GitHub staff to assist here.
>@MikeMcQuaid @lildude @brandonblack - there's an area of staff tools devtools/samples I think that allows you to run a search query and it batches up the first ~1000 files. These are useful for testing new heuristics. Unfortunately the jobs don't seem to be running any longer and I didn't manage to figure out what was going on before I left. Any chance one of you could take a look?

Ping @MikeMcQuaid @lildude @brandonblack - this PR is pending for quite a while now. Would you be so kind to find a bit of time to look at the search query? AFAICT we only need to testing this PR against your GitHub corpus.

**Comment by wilzbach** [comment]
> Ping @MikeMcQuaid @lildude @brandonblack - this PR is pending for quite a while now. Would you be so kind to find a bit of time to look at the search query? AFAICT we only need to testing this PR against your GitHub corpus.

Friendly re-ping ;-)

**Comment by Alhadis** [comment]
Polite reminder that we all have lives, and Mr McQuaid is responsible for maintaining the biggest (and also the best) package manager in the solar system.

I'll have a look into this later.

**Comment by lildude** [comment]
Sorry about the delay here. I started looking into the failing job of your last ping @wilzbach  but then my own `$job` stole my attention. I'll try find some time this week to see if I work out why the job is failing. I have a suspicion I know what the problem is.

**Comment by MartinNowak** [comment]
Build a scraper myself and improved all the existing heuristics for the `.d` file ending.

**Comment by MartinNowak** [comment]
Can we get a fresh review on this @Alhadis, @brandonblack?

File: lib/linguist/heuristics.rb [hunk 2]
```diff
@@ -125,11 +125,18 @@ def call(data)
...
-      elsif /^((dtrace:::)?BEGIN|provider |#pragma (D (option|attributes)|ident)\s)/.match(data)
+      # see http://dtrace.org/guide/chp-prog.html, http://dtrace.org/guide/chp-profile.html, http://dtrace.org/guide/chp-opt.html
+      elsif /^(\w+:\w*:\w*:\w*|BEGIN|END|provider\s+|(tick|profile)-\w+\s+{[^}]*}|#pragma\s+D\s+(option|attributes|depends_on)\s|#pragma\s+ident\s)/.match(data)
         Language["DTrace"]
-      elsif /(\/.*:( .* \\)$| : \\$|^ : |: \\$)/.match(data)
+      # path/target : dependency \
+      # target : \
+      #  : dependency
+      # path/file.ext1 : some/path/../file.ext2
+      elsif /([\/\\].*:\s+.*\s\\$|: \\$|^ : |^[\w\s\/\\.]+\w+\.\w+\s*:\s+[\w\s\/\\.]+\w+\.\w+)/.match(data)
```
Comment (MartinNowak) [inline lib/linguist/heuristics.rb]: The last regex part for `path/file.ext1 : some/path/../file.ext2` is specific to match only files with extensions b/c `Symbol : Base` is commonly used in the other languages as well.

**Comment by MartinNowak** [comment]
Already rotting since 2 month.
Improves 3 languages, thoroughly tested, documented regexes, shouldn't require much effort to review.
@pchaigno, @brandonblack

**Comment by larsbrinkhoff** [comment]
I'm still in favour of this change.  Good work.

Patience and the occasional ping may be required.

**Comment by MartinNowak** [comment]
> I meant to get back to this earlier ... sorry.

No worries, fairly busy myself :).

File: Rakefile [hunk 3]
```diff
@@ -4,6 +4,7 @@ require 'rake/testtask'
 require 'yaml'
 require 'yajl'
 require 'open-uri'
+require 'json'
```
Comment (lildude) [inline Rakefile]: I can't see any reason for this addition.  Is there a reason for this?  If not, can you please remove it?

Comment (MartinNowak) [inline Rakefile]: The rake diff task fails w/o it, b/c it uses `Hash.to_json`. Apparently I was the first to use that in a while. Also see the separate commit message https://github.com/github/linguist/pull/3145/commits/7980c3dc430b4100eee015c23ba5c12bfe027694.
I could for sure separate that out into a separate PR if wanted, but it seems like a tiny fix on-the-go.

Comment (MartinNowak) [inline Rakefile]: The rake diff task fails w/o it, b/c it uses Hash.to_json. Apparently I was the first to use that in a while. Also see the separate commit message 7980c3d.
I could for sure separate that out into a separate PR if wanted, but it seems like a tiny fix on-the-go.
@lildude

Comment (lildude) [inline Rakefile]: > I could for sure separate that out into a separate PR if wanted, but it seems like a tiny fix on-the-go.

Nah, that's good enough explanation for me to do it here 😄.

**Comment by MartinNowak** [comment]
[202 Accepted Cat](https://http.cat/202)

We made it :checkered_flag:!
You guys should add reactions to commits and merge messages :wink:.
